src/user_database.ts

```diff
@@ -84,13 +87,18 @@ export class PGNodeUserDatabase implements UserDatabase {
     }
   }
 
-  async query<R>(sql: string, ...params: any[]): Promise<R[]> {
+  async queryFunction<C extends UserDatabaseClient, R, T extends unknown[]>(func: UserDatabaseQuery<C, R, T>, ...args: T): Promise<R> {
+    const client: PoolClient = await this.pool.connect();
+    return func(client as C, ...args);
```

Comment:
I guess this question applies to all dbClient: how do we enforce queryFunction to be read-only?

Reply:
I am thinking that would depend on the underlying DB api, perhaps if you said transaction, read only, no isolation you would get read-only without hurting performance, but I don't know...

Reply:
I think if we wrap the queryFunction in a transaction and tell Postgres that this is a `read-only` transaction, PG will throw an error if you try to update anything. Then if we use transaction, we should just use read-committed isolation (the default one in PG).

Reply:
I just found out that you can create a read-only Prisma client: https://github.com/prisma/prisma-client-extensions/tree/main/readonly-client

Reply:
Agree--do it in a client-specific way if possible, and if we need to follow back on Postgres's `SET TRANSACTION READ ONLY`.